libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,315 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+    if (ERROR_SUCCESS != Status) {
+        LogError("Driver not installed");
+        return Status;
+    }
+
+    
+    PWNBD_CONNECTION_LIST ConnList = NULL;
+    DWORD err = HardRemoveAllDevices();
+    if (err) {
+        return err;
+    }
+    
+    Status = WnbdRemoveDriver(AdapterDevHandle, &DevInfoData, RebootRequired);
+
+    if (INVALID_HANDLE_VALUE != AdapterDevHandle) {
+        SetupDiDestroyDeviceInfoList(AdapterDevHandle);
+    }
+
+    if (ERROR_SUCCESS != Status) {
+        LogError(
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+    }
+
+    return Status;
+}
+
+DWORD InstallDriver(CHAR* FileNameBuf, HDEVINFO* AdapterDevHandle, PBOOL RebootRequired)
+{
+    CHAR ClassName[MAX_CLASS_NAME_LEN];
+    GUID ClassGuid;
+    SP_DEVINFO_DATA DevInfoData;
+    SP_DEVINSTALL_PARAMS_A InstallParams;
+    DWORD Status = 0;
+
+    if (!SetupDiGetINFClassA(FileNameBuf, &ClassGuid, ClassName, MAX_CLASS_NAME_LEN, 0)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (INVALID_HANDLE_VALUE == (*AdapterDevHandle = SetupDiCreateDeviceInfoList(&ClassGuid, 0))) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    DevInfoData.cbSize = sizeof DevInfoData;
+    if (!SetupDiCreateDeviceInfoA(*AdapterDevHandle, ClassName, &ClassGuid, 0, 0, DICD_GENERATE_ID, &DevInfoData)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (!SetupDiSetDeviceRegistryPropertyA(*AdapterDevHandle, &DevInfoData,
+                                           SPDRP_HARDWAREID, (PBYTE)WNBD_HARDWAREID, WNBD_HARDWAREID_LEN)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (!SetupDiCallClassInstaller(DIF_REGISTERDEVICE, *AdapterDevHandle, &DevInfoData)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    InstallParams.cbSize = sizeof InstallParams;
+    if (!SetupDiGetDeviceInstallParamsA(*AdapterDevHandle, &DevInfoData, &InstallParams)) {
+        Status = GetLastError();
+        goto exit;
+    }
+
+    if (0 != (InstallParams.Flags & (DI_NEEDREBOOT | DI_NEEDRESTART))) {
+        *RebootRequired = TRUE;
+    }
+
+exit:
+    return Status;
+}
+
+DWORD WnbdInstallDriver(PSTR FileName, PBOOL RebootRequired)
+{
+    CHAR FullFileName[MAX_PATH];
+    DWORD Status = ERROR_SUCCESS;
+    HDEVINFO AdapterDevHandle = INVALID_HANDLE_VALUE;
+    SP_DEVINFO_DATA DevInfoData;
+    DevInfoData.cbSize = sizeof DevInfoData;
+
+    if (0 == GetFullPathNameA(FileName, MAX_PATH, FullFileName, 0)) {
+        Status = GetLastError();
+        LogError(
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+    if (FALSE == PathFileExistsA(FullFileName)) {
+        LogError("File: %s, does not exist", FullFileName);
+        Status = ERROR_FILE_NOT_FOUND;
+        goto exit;
+    }
+    if (ERROR_SUCCESS == FindDevice(&AdapterDevHandle, &DevInfoData)) {
+        LogError("Driver already installed");
```

Comment:
This is a bit odd, we're checking if there's an existing WNBD device, not an existing driver. What if there's no WNBD device but there's an existing driver?

Reply:
Not necessarily, our own tool might successfully install the driver but fail to create the device (e.g. due to an error).